gordo_components/dataset/datasets.py

```diff
@@ -86,42 +72,161 @@ def __init__(
+            See :py:func::`pandas.core.resample.Resampler#aggregate` for more
+            information on possible aggregation methods.
+        kwargs: dict
+            kwargs to pass to :class:`RawTimeSeriesDataset`
+        """
+        self.resolution = resolution
+        self.aggregation_methods = aggregation_methods
+        super().__init__(**kwargs)
+
+    def get_data(self) -> Tuple[pd.DataFrame, Optional[pd.DataFrame]]:
+        series_iter: Iterable[pd.Series] = self.data_provider.load_series(
+            from_ts=self.from_ts,
+            to_ts=self.to_ts,
+            tag_list=list(set(self.tag_list + self.target_tag_list)),
+        )
+        data: pd.DataFrame = self.join_timeseries(
+            series_iter,
+            self.from_ts,
+            self.to_ts,
+            self.resolution,
+            aggregation_methods=self.aggregation_methods,
+        )
+        if self.row_filter:
+            data = pandas_filter_rows(data, self.row_filter)
+        return self._X_y_from_dataframe(data)
+
+    def get_metadata(self):
+        metadata = super().get_metadata()
+        metadata.update(
+            {
+                "aggregation-methods": self.aggregation_methods,
+                "resolution": self.resolution,
+            }
+        )
+        return metadata
+
+
+class ShiftedTimeSeriesDataset(TimeSeriesDataset):
+    def __init__(self, shift: int = 1, *args, **kwargs):
+        """
+        Return the same data as given by :class:`TimeSeriesDataset`
+        but offset the ``y`` such that the Nth index of ``y`` relates to the
+        datetime in index ``N - shift`` index of ``X``.
+
+        Used if you want to train a model to predict future time step values.
+
+        Parameters
+        ----------
+        shift: int
+            Amount to offset the target column.
+        args: tuple
+            Any arguments to supply which are passed onto :class:`TimeSeriesDataset`
+        kwargs: dict
+            Any keyword arguments to be passed into :class:`TimeSeriesDataset`
+
+        Examples
+        --------
+        If the ``resolution`` is 10 minutes and a provided ``shift`` is 1, then
+        the target value(s) for any sample of ``X`` would have the target value(s)
+        10 minutes into the future.
+        """
+        self.shift = shift
+        super().__init__(*args, **kwargs)
+
+    @property
+    def shift(self) -> int:
+        return self._shift
+
+    @shift.setter
+    def shift(self, value: int):
+        self._shift = int(value)
+
+    def get_metadata(self):
+        metadata = super().get_metadata()
+        metadata.update({"shift": self.shift})
+        return metadata
+
+    def get_data(self) -> Tuple[pd.DataFrame, Optional[pd.DataFrame]]:
+        X, y = super().get_data()
+        if self.shift > 0:
+            return X[: -self.shift], y[self.shift :] if y is not None else None
+        else:
+            return X[self.shift :], y[: -self.shift] if y is not None else None
```

Comment:
Exciting stuff. One question: if y is none, then we just return None, right? Even if X is not none. This makes sense, but perhaps we should log an ERROR and/or throw an exception here, since you have a "shifted" timeseries set, but have not provided a y, which doesn't make sense?